### PR TITLE
[None][chore] Split TorchSampler feature code out of sampler.py

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Description

`sampler.py` had grown to 4568 lines, with `TorchSampler` alone accounting for ~2600 of them.
This PR splits the self-contained pieces into their own modules, following the
`<Feature>Store` / `<Feature>Handler` convention the package already uses for beam search,
penalties, token bans and top-p decay.

Pure code motion — no behavior change. `sampler.py`: **4568 -> 2886 lines**.

### New modules

| Module | Contents |
|---|---|
| `two_model_spec_dec.py` | Draft-token verification for two-model speculation (greedy / tree / rejection sampling). Only reached when the drafter runs as a separate engine; one-model speculation verifies via `SpecSampler`. Also absorbs `get_rejected_indices` / `sample_rejected` from `ops/vanilla.py`, whose only consumer this is. |
| `seed_manager.py` | `_SeedManager` — per-slot Philox `(seed, offset)` for `SamplingParams.seed`. |
| `sampler_features.py` | Small features that do not warrant a module of their own: step indexing (`_PackedStepIndexer` / `_UnpackedStepIndexer`), host-side stop criteria, logit adjustments (embedding bias, d2t, fused greedy kernel) and the async-D2H machinery (`AsyncWorkerMixin`, `SamplerEvent`, side-stream copier). |

### Moved into existing modules

| From | To |
|---|---|
| `_CachingRequestGrouper` + `RequestGroup*` | `sampler_strategy.py` — the grouper buckets by `Strategy`, and the grouped samplers that consume its rows already live there |
| logprobs sizing + per-step device gather | `logprobs.py` (`LogProbsHandler`) |
| beam-search metadata assembly (~60 lines) | `BeamSearchHandler.build_metadata`, mirroring `TopPDecayHandler.build_metadata` |
| `convert_logprobs_tensor_to_list` | `beam_search.py` — defined in `logprobs.py` but called only from `finalize_beam` |

### Dependency graph

Several types were sunk or re-homed so imports only ever point downward:

- `RequestSeeds`, `top_p_decay_active`, `_BatchedSamplingResult` -> `sampler_common.py`
- `TopPDecayMetadata` -> `top_p_decay.py`, so the feature owns its own metadata type
  (matching how `beam_search.py` owns `BeamSearchMetadata`)

This drops `seed_manager` and `top_p_decay` from the deepest layer to the shallowest, and lets
`logprobs.py` and `beam_search.py` drop their import-cycle workarounds (a function-local import
and a `TYPE_CHECKING` + string-annotation dance respectively).

There are no import cycles. The two remaining back-references to `sampler.py` are
`TYPE_CHECKING`-only, for a handler's annotation of its owning `TorchSampler`.

### Docs

Module docstrings were audited against actual contents. Fixed: `logprobs.py` still claimed
sizing lived in `TorchSampler`; `beam_search.py` referenced a `sampling_utils` module that no
longer exists. Added a module docstring to `sampler.py` (noting `TRTLLMSampler` is slated for
removal in 1.4, after which this module should be reorganized) and a layout diagram to
`__init__.py`.

### mypy

The three new modules are added to the explicit `[tool.mypy] files` list in `pyproject.toml`,
which is the single source of truth for the type-check file set. Code moved out of the
already-listed `sampler.py` would otherwise silently lose its mypy coverage.

## Test Coverage

No new tests — this is code motion, and the existing suites cover the moved code. One test
call path was updated: `test_draft_token_tree_verification.py` now reaches
`_process_draft_tokens_tree` through `torch_sampler._two_model_spec_dec`.

Relevant existing tests:

- `tests/unittest/_torch/sampler/test_torch_sampler.py`
- `tests/unittest/_torch/sampler/test_logits_logprobs.py`
- `tests/unittest/_torch/sampler/test_beam_search.py`
- `tests/unittest/_torch/sampler/test_beam_search_speculative_d2h.py`
- `tests/unittest/_torch/sampler/test_token_ban.py`, `test_penalties.py`, `test_best_of_n.py`
- `tests/unittest/_torch/speculative/hw_agnostic/test_torch_rejection_sampling.py`
- `tests/unittest/_torch/speculative/hw_agnostic/test_draft_token_tree_verification.py`

Ran locally on B200 (1103 passed). Caveat on the local run: it used the `1.3.0rc24` release
container's compiled bindings with this branch's sampler package overlaid, since the change is
Python-only. Its `_torch/pyexecutor/` is byte-identical to this branch's base, but the C++
bindings are older, so 5 `test_vbws_*` cases in `test_beam_search.py` fail there on
`get_beam_width_by_iter` returning garbage. Those tests import nothing from the sampler package
and this PR touches neither `llm_request.py` nor `test_beam_search.py`. Relying on CI for a
matched-build run.

## PR Checklist

- [x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- Refactors `TorchSampler` into focused modules for sampler features, seed management, speculative decoding, logprobs, beam search, and request grouping.
- Reduces `sampler.py` from 4,568 to 2,886 lines.
- Fixes embedding-bias mask indexing during speculative decoding.
- Preserves intended sampling behavior and updates shared imports, documentation, package layout, and mypy configuration.
- Review focus: verify import boundaries, seeded-request reproducibility, speculative-decoding behavior, asynchronous copy ordering, relocated symbol compatibility, and error handling.
- No test-list files changed.

## QA Engineer Review

- Modifies `test_draft_token_tree_verification.py` to call `_process_draft_tokens_tree` through `TorchSampler._two_model_spec_dec`.
- No test functions were added or removed.
- No `tests/integration/test_lists/` coverage entry changed.
- Verdict: needs follow-up.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
Thanks @coderabbitai — both findings point at real code, but neither is introduced by this PR, and one of them is a false positive. Details below.

### 1. `apply_embedding_bias` mask offset — **valid, fixed in 5819918**

Confirmed and fixed. `i` is the request index but the mask indexes rows of `logits`, and a request contributes `steps` rows (`> 1` under speculative decoding). Your repro is right: two bias-carrying requests with `steps == 2` mark rows `0,1` then `1,2`, giving 3 `True` entries against 4 `bias_gather_indices`, so `logits[mask] += biases` raises on the shape mismatch. A biased request near the end of the batch can also index past `logits.size(0)`.

Fixed by tracking a running row offset, as suggested. `bias_gather_indices` was already built in row order, so only the mask was wrong, and the single-step path is bit-identical.

Note this is **pre-existing on `main`**, not introduced here — the function is byte-identical to `TorchSampler._apply_embedding_bias` before it moved to `sampler_features.py` (verified by AST comparison against the base commit `0eeda343df`). There is currently no test coverage for embedding bias anywhere under `tests/unittest/_torch/sampler/`, which is why it survived; a regression test needs a new fixture rather than an extension of an existing one, so it is not in this PR.

### 2. `sample_rejected` indexing `target_probs` — **false positive, no change made**

`target_probs` here is **2-D `[steps, vocab_size]`**, with no leading batch dimension. It is assigned in `sampler.py` as a slice of the per-group softmax:

```python
requests[req_idx].py_target_probs = group_softmax_cuda[current_offset:next_offset]
```

So `target_probs[num_accepted]` already selects one step's probability vector `[vocab_size]`, which is what `torch.multinomial` needs. The suggested `target_probs[0, num_accepted]` would index a **scalar** and break the call.

The `.squeeze(0)` in `get_rejected_indices` is not evidence of a batch dimension: `torch.squeeze(0)` is a no-op unless `dim 0` has extent 1, so for the usual `steps > 1` case it changes nothing, and the two functions are consistent. Left unchanged.

This code is also byte-identical to `main` (it moved here from `ops/vanilla.py`, whose only consumer is this module).

File: tensorrt_llm/_torch/pyexecutor/sampler/two_model_spec_dec.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    target_probs: torch.Tensor,
+    generator: torch.Generator,
+    num_accepted: int,
+) -> int:
+    last_draft = draft_probs[num_accepted]
+    last_target = target_probs[num_accepted]
+    new = last_target - last_draft
+    new = torch.where(new > 0, new, 0.0)
+    new_token = torch.multinomial(new, num_samples=1, generator=generator).squeeze(-1)
+    return cast(int, new_token.item())
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/two_model_spec_dec.py:79]: Not applying this one — `target_probs` has no leading batch dimension here.

It is assigned in `sampler.py` as a slice of the per-group softmax, so its shape is `[steps, vocab_size]`:

```python
requests[req_idx].py_target_probs = group_softmax_cuda[current_offset:next_offset]
```

`target_probs[num_accepted]` therefore already selects one step's probability vector `[vocab_size]`, which is what `torch.multinomial` expects. The suggested `target_probs[0, num_accepted]` would index a scalar and break the call.

The `.squeeze(0)` in `get_rejected_indices` is not evidence of a batch dimension: `torch.squeeze(0)` is a no-op unless dim 0 has extent 1, so for the usual `steps > 1` case it changes nothing and the two functions are consistent.

This code is also unchanged by this PR — it moved verbatim from `ops/vanilla.py` (verified by AST comparison against the base commit `0eeda343df`).

Acknowledged as incorrect in https://github.com/NVIDIA/TensorRT-LLM/pull/17890#issuecomment-5327178489. Resolving.

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/sampler/beam_search.py [hunk 2]
```diff
@@ -12,35 +12,41 @@
...
 # limitations under the License.
 
-"""PyTorch-native beam-search sampling kernels.
+"""Beam search for ``TorchSampler``.
 
-The candidate-selection, beam-expansion, and candidate-beams-array (CBA)
-beam-search logic for TorchSampler, split out of ``ops.vanilla``. Pure tensor
-functions plus their metadata dataclasses; no dependency on the
-``sampling_utils`` interface.
+The candidate-selection, beam-expansion and candidate-beams-array (CBA) logic,
```
Comment (tongyuantongyu) [inline tensorrt_llm/_torch/pyexecutor/sampler/beam_search.py]: “originally split out of ops.vanilla” describes repository history that does not help readers understand the current module. Please describe only its present responsibility, such as: “This module contains the candidate-selection, beam-expansion, and candidate-beams-array logic, along with the metadata types those operations consume.”

File: tensorrt_llm/_torch/pyexecutor/sampler/seed_manager.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    """Per-sequence-slot RNG state implementing ``SamplingParams.seed``.
+
+    A seeded request must produce the same tokens regardless of which other
+    requests share its batch, so its RNG stream cannot come from a single
+    batch-wide ``torch.Generator`` whose state advances by the batch's total
+    draw count. Instead each row is sampled with an explicit Philox
+    ``(seed, offset)`` pair: ``seed`` is the user's seed and ``offset`` is that
+    request's own running draw count, both indexed by sequence slot.
+
+    Requests without a seed fall back to ``global_seed`` and share the same
```
Comment (tongyuantongyu) [inline tensorrt_llm/_torch/pyexecutor/sampler/seed_manager.py:44]: “does not reproduce the pre-existing batch-wide generator stream” assumes knowledge of an earlier implementation. Please document the observable current semantics instead: unseeded requests use global_seed with per-slot offsets and therefore are deterministic per slot, but their token streams may differ when batching or slot assignment changes.

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler_common.py [hunk 4]
```diff
@@ -14,19 +14,35 @@
...
+* the per-request queries that read an ``LlmRequest``'s sampling config into
+  :class:`UtilsSamplingParams`, plus the predicates over it
+  (``top_p_decay_active``, ``_request_sampling_params_cachable``);
+* the shared step/beam index constants and beam-width accessors;
+* tensor helpers (``int_tensor``, ``add_token``);
+* plain data types passed *between* modules that no single feature owns --
+  :class:`RequestSeeds` (seed manager -> strategy impls) and
+  :class:`_BatchedSamplingResult` (sampler -> log-probs).
+
+That last group is also how a type-level import cycle gets broken: when a
```
Comment (tongyuantongyu) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler_common.py]: “gets broken,” “type is sunk here,” and “imported backwards” describe a refactoring operation rather than the current dependency rule. Please phrase this as a standing architectural constraint: shared cross-feature types belong in this base module so lower-level modules never need to import higher-level modules; feature-owned types remain with their feature.

**Comment by zhaoyangwang-nvidia** [comment]
The `test_spec_dec_stats_pairing` failure from the previous run is fixed (`f8ff7ed`). The remaining failure is the DeepSeek-V3-Lite host-perf sanity case, and I believe it is noise rather than a regression from this PR. Evidence:

**1. The failure shape is a stall, not a slowdown.**

| metric | value / baseline | ratio |
|---|---|---|
| `d_p99_ttft` | 2427.12 / 112.23 | **21.6x** |
| `d_p99_e2el` | 3751.29 / 1355.07 | 2.77x |
| `d_mean_ttft` | 210.18 / 105.03 | 2.00x |
| `d_mean_itl` | 5.24 / 4.65 | 1.13x |
| `d_token_throughput` | 1325.06 / 1554.54 | 0.85x |

A host-overhead regression raises steady-state cost uniformly. Here `p99_ttft` is 21.6x while `mean_itl` moves 13% — i.e. a single request waited ~2.4 s for its first token, which then drags mean TTFT and the aggregate throughput numbers. That is the signature of a one-off scheduling/contention event on the node, not of extra Python work per iteration.

**2. This config does not exercise the changed code.**

`v3lite_fp8_bs8_128_256` runs plain greedy decode: no `logprobs`, no `embedding_bias`, no beam search, no speculative decoding. The handlers this PR introduces for those features are constructed once and never entered on this path.

**3. The code on the hot path is byte-identical to `main`.**

Verified by AST comparison against `origin/main`:

- `fast_greedy_sample_kernel` — identical
- `apply_d2t` — identical
- `_process_logprobs` — identical after the mechanical `self.X` → `self._sampler.X` rewrite (0 diff lines)

The only functions that differ are `apply_embedding_bias` (two bug fixes in `1b295a4`/`6264cf3`, on a path this config never takes). The rest of the PR is code motion: no loop, allocation, `.item()` or `.tolist()` was added. The added indirection is one attribute lookup per handler call, on paths that are per-step at most.

Re-running to confirm. If it reproduces deterministically I will profile it rather than dismiss it.

File: tensorrt_llm/_torch/pyexecutor/sampler/beam_search.py (hunk 2 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/beam_search.py]: Done in `6db7cef`. Dropped the "originally split out of `ops.vanilla`" clause; the paragraph now just states what the module holds.

File: tensorrt_llm/_torch/pyexecutor/sampler/seed_manager.py (hunk 3 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/seed_manager.py:44]: Done in `6db7cef`. Now states the semantics directly: unseeded requests are deterministic for a given sequence slot, but their token stream is not stable across runs that batch them differently or assign a different slot.

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler_common.py (hunk 4 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler_common.py]: Done in `6db7cef`. Rephrased as a standing rule: types shared across features live in this base module so a lower layer never imports a higher one; feature-owned types stay with their feature.

**Comment by zhaoyangwang-nvidia** [comment]
Thanks — fair point, and the phrasing was mine to fix. Corrected in `6db7cef` (comments only, no code change). I also swept the rest of the PR's added lines for the same pattern and found one more in `sampler.py`, where the note about the planned `TRTLLMSampler` removal said the remaining classes "no longer need to share a file"; it now describes the end state directly.
